<!-- 
  @author Simone Cottiga
  @email simone.cottiga@phd.units.it

  @author Davide Panarella
  @email davidepanarella.ts@gmail.com
 -->

<launch>
  <!-- gridmap arguments -->
  <arg name="/use_sim_time" default="true"/>
  <arg name="gridmap_config_file" default="archimede_gridmap.yaml"/>

  <!-- archimede_gazebo simulations arguments -->
  <arg name="include_gazebo" default="false"/>
  <arg name="include_teleop" default="false"/>
  <arg name="rocker_differential" default="false"/>
  <arg name="world_name" default="tilted_floor_walls_and_obstacles.world"/>
  <arg name="p38_env" default="/home/ros/p38Env/bin/python3.8"/>
  <arg name="add_velodyneHDL32E" default="true"/>

  <!-- include_color_detection arguments -->
  <arg name="include_color_detection" default="false"/>

  <!-- link_state_to_odom node argument -->
  <arg name="include_linkstates2odom" default="false"/>

  <!-- play bag arguments -->
  <arg name="include_bag" default="false"/>  
  <arg name="bag_file" default="/media/ros/142F8F8C794D9A16/test_filter_bag.bag"/>

  <!-- grid_map stuffs -->
  <rosparam command="load" file="$(find robot4ws_mapping)/config/$(arg gridmap_config_file)" ns="gridmap"/>

  <node name="cloud_filter_node" pkg="robot4ws_mapping" type="cloud_filter_node" output="screen"/>
  <!--<node name="p2g_listener" pkg="robot4ws_mapping" type="gridmap_node" output="screen"/> -->
  <node name="multi_layer_map_node" pkg="robot4ws_mapping" type="multi_layer_map_node" output="screen"/>
  <node name="elevation_node" pkg="robot4ws_mapping" type="elevation_node" output="screen"/>
  <node name="surface_orientation_node" pkg="robot4ws_mapping" type="surface_orientation_node" output="screen"/>
  <node name="obstacles_node" pkg="robot4ws_mapping" type="obstacles_node" output="screen" />

  <node name="rviz" pkg="rviz" type="rviz" args="-d $(find robot4ws_mapping)/rviz/grid_map.rviz" />

  <!-- include archimede_gazebo -->
  <group if="$(arg include_gazebo)">
    <group unless="$(arg include_teleop)">
      <include file="$(find robot4ws_simulations)/launch/archimede_gazebo_simulation_with_kinematics.launch">
        <arg name="rocker_differential" value="$(arg rocker_differential)"/>
        <arg name="world_name" value="$(arg world_name)"/>
        <arg name="p38_env" value="$(arg p38_env)"/>
        <arg name="add_velodyneHDL32E" value="$(arg add_velodyneHDL32E)"/>
      </include>
    </group>
    <group if="$(arg include_teleop)">
      <include file="$(find robot4ws_simulations)/launch/archimede_gazebo_simulation_remote_controlled.launch">
        <arg name="rocker_differential" value="$(arg rocker_differential)"/>
        <arg name="world_name" value="$(arg world_name)"/>
        <arg name="p38_env" value="$(arg p38_env)"/>
        <arg name="add_velodyneHDL32E" value="$(arg add_velodyneHDL32E)"/>
      </include>
    </group>
  </group>

  <!-- include link_state_to_odom -->
  <group if="$(arg include_linkstates2odom)">
    <include file="$(find robot4ws_mapping)/launch/link_state_to_odom.launch"/>
  </group>

  <!-- include play bag -->
  <group if="$(arg include_bag)">
    <node pkg="rosbag" type="play" name="rosbag_play" args="$(arg bag_file)" output="screen"/>
  </group>

  <!-- include link_state_to_odom -->
  <group if="$(arg include_color_detection)">
    <include file="$(find robot4ws_perception)/launch/color_detection.launch"/>
  </group>

</launch>

